diff --git a/builtin/archive.c b/builtin/archive.c
index a1e3b94..fe1034e 100644
--- a/builtin/archive.c
+++ b/builtin/archive.c
@@ -15,6 +15,12 @@ static void create_output_file(const char *output_file)
 	int output_fd = open(output_file, O_CREAT | O_WRONLY | O_TRUNC, 0666);
 	if (output_fd < 0)
 		die_errno(_("could not create archive file '%s'"), output_file);
+#ifdef __MVS__
+ #if (__CHARSET_LIB == 1)
+        if (setbinaryfd(output_fd))
+		die_errno(_("could not tag archive file '%s'"), output_file);
+ #endif
+#endif
 	if (output_fd != 1) {
 		if (dup2(output_fd, 1) < 0)
 			die_errno(_("could not redirect output"));
