diff --git a/builtin.h b/builtin.h
index 6b95006..3b7dd57 100644
--- a/builtin.h
+++ b/builtin.h
@@ -143,4 +143,8 @@ extern int cmd_show_ref(int argc, const char **argv, const char *prefix);
 extern int cmd_pack_refs(int argc, const char **argv, const char *prefix);
 extern int cmd_replace(int argc, const char **argv, const char *prefix);
 
+#ifdef __MVS__
+  extern int setbinaryfd(int);
+#endif
+
 #endif
