diff --git a/config.mak.uname b/config.mak.uname
index d63629f..dd1d70f 100644
--- a/config.mak.uname
+++ b/config.mak.uname
@@ -622,6 +622,27 @@ ifeq ($(uname_S),NONSTOP_KERNEL)
 	SANE_TOOL_PATH = /usr/coreutils/bin:/usr/local/bin
 	SHELL_PATH = /usr/coreutils/bin/bash
 endif
+
+ifeq ($(uname_S),OS/390)
+  V = 2
+  PERL_PATH = perl
+  PYTHON_PATH = python
+  HAVE_ALLOCA_H = YesPlease
+  NO_SYS_POLL_H = YesPlease
+  	BLK_SHA1 = YesPlease
+  NO_STRCASESTR = YesPlease
+  NO_REGEX = NeedsStartEnd
+  NO_NSEC = YesPlease
+  NO_STRLCPY = YesPlease
+  NO_MKDTEMP = YesPlease
+  NO_MEMMEM = YesPlease
+  NO_GECOS_IN_PWENT = YesPlease
+  NEEDS_SSL_WITH_CRYPTO = YesPlease
+  NEEDS_MODE_TRANSLATION = YesPlease
+  HAVE_STRINGS_H = YesPlease
+  PTHREAD_LIBS =
+endif
+
 ifeq ($(uname_S),MINGW)
 	pathsep = ;
 	HAVE_ALLOCA_H = YesPlease
