diff --git a/config.mak.uname b/config.mak.uname
index d63629f..70bd89c 100644
--- a/config.mak.uname
+++ b/config.mak.uname
@@ -362,7 +362,7 @@ ifeq ($(uname_S),IRIX)
 	NO_MMAP = YesPlease
 	NO_REGEX = YesPlease
 	SNPRINTF_RETURNS_BOGUS = YesPlease
-	SHELL_PATH = /usr/gnu/bin/bash
+	SHELL_PATH = "/bin/env bash"
 	NEEDS_LIBGEN = YesPlease
 endif
 ifeq ($(uname_S),IRIX64)
@@ -622,6 +622,27 @@ ifeq ($(uname_S),NONSTOP_KERNEL)
 	SANE_TOOL_PATH = /usr/coreutils/bin:/usr/local/bin
 	SHELL_PATH = /usr/coreutils/bin/bash
 endif
+
+ifeq ($(uname_S),OS/390)
+  V = 2
+	PERL_PATH = /bin/env perl
+	SHELL_PATH = bash
+	SHELL_PATH_FOR_SCRIPTS = /bin/env bash
+	PYTHON_PATH = /bin/env python3
+  NO_SYS_POLL_H = YesPlease
+  NO_STRCASESTR = YesPlease
+  NO_REGEX = YesPlease
+  NO_MMAP = YesPlease
+  NO_NSEC = YesPlease
+  NO_STRLCPY = YesPlease
+  NO_MKDTEMP = YesPlease
+  NO_MEMMEM = YesPlease
+  NO_GECOS_IN_PWENT = YesPlease
+  HAVE_STRINGS_H = YesPlease
+  PTHREAD_LIBS =
+  NEEDS_MODE_TRANSLATION = YesPlease
+endif
+
 ifeq ($(uname_S),MINGW)
 	pathsep = ;
 	HAVE_ALLOCA_H = YesPlease
