diff --git a/pager.c b/pager.c
index b66bbff..f0e90fd 100644
--- a/pager.c
+++ b/pager.c
@@ -66,7 +66,12 @@ const char *git_pager(int stdout_is_tty)
 	if (!*pager || !strcmp(pager, "cat"))
 		pager = NULL;
 
+#ifdef __MVS__
+  // The location pointed to by pager gets overwritten by z/OS LE, duplicate it as a workaround
+	return xstrdup(pager);
+#else
 	return pager;
+#endif
 }
 
 static void setup_pager_env(struct strvec *env)
