diff --git a/wrapper.c b/wrapper.c
index 26db215..9bec8c0 100644
--- a/wrapper.c
+++ b/wrapper.c
@@ -418,6 +418,42 @@ FILE *fopen_for_writing(const char *path)
 	return ret;
 }
 
+#ifdef __MVS__
+# if (__CHARSET_LIB == 1)
+#   include <stdio.h>
+#   include <stdlib.h>
+
+    static int setccsid(int fd, int ccsid) 
+    {
+      attrib_t attr;
+      int rc;
+
+      memset(&attr, 0, sizeof(attr));
+      attr.att_filetagchg = 1;
+      attr.att_filetag.ft_ccsid = ccsid;
+      attr.att_filetag.ft_txtflag = 1;
+
+      rc = __fchattr(fd, &attr, sizeof(attr));
+      return rc;
+    }
+
+    static int mkstempascii(const char* template) 
+    {
+	int fd = mkstemp(template);
+	if (fd) {
+		/* MSF - TBD - do not ALWAYS tag this as ASCII - perhaps usage may require more finesse */
+		if (setccsid(fd, 819)) {
+			die_errno("mkstemp failed");
+			fd = 0;
+		}
+	}
+	return fd;
+    }
+
+#   define mkstemp mkstempascii
+  #endif
+#endif
+		
 int xmkstemp(char *template)
 {
 	int fd;
