diff --git a/builtin/archive.c b/builtin/archive.c
index 13ea730..827bc18 100644
--- a/builtin/archive.c
+++ b/builtin/archive.c
@@ -8,10 +8,17 @@
 #include "transport.h"
 #include "parse-options.h"
 #include "pkt-line.h"
+#ifdef __MVS__
+#include <zos.h>
+#endif
 
 static void create_output_file(const char *output_file)
 {
 	int output_fd = xopen(output_file, O_CREAT | O_WRONLY | O_TRUNC, 0666);
+#ifdef __MVS__
+        if (__setfdbinary(output_fd))
+		die_errno(_("could not tag archive file '%s'"), output_file);
+#endif
 	if (output_fd != 1) {
 		if (dup2(output_fd, 1) < 0)
 			die_errno(_("could not redirect output"));
