diff --git a/builtin.h b/builtin.h
index f7b166b334..c7896cda4b 100644
--- a/builtin.h
+++ b/builtin.h
@@ -253,5 +253,7 @@ int cmd_verify_pack(int argc, const char **argv, const char *prefix, struct repo
 int cmd_show_ref(int argc, const char **argv, const char *prefix, struct repository *repo);
 int cmd_pack_refs(int argc, const char **argv, const char *prefix, struct repository *repo);
 int cmd_replace(int argc, const char **argv, const char *prefix, struct repository *repo);
-
+#ifdef __MVS__
+  extern int setbinaryfd(int);
+#endif
 #endif
