diff --git a/cache.h b/cache.h
index bdedb87..2c4ba41 100644
--- a/cache.h
+++ b/cache.h
@@ -8,6 +8,21 @@
 #include "object.h"
 #include "statinfo.h"
 
+// Cannot include dir.h for this because there is a conflict with dir_entry struct
+#if defined(DT_UNKNOWN) && !defined(NO_D_TYPE_IN_DIRENT)
+#define DTYPE(de)	((de)->d_type)
+#else
+#undef DT_UNKNOWN
+#undef DT_DIR
+#undef DT_REG
+#undef DT_LNK
+#define DT_UNKNOWN	0
+#define DT_DIR		1
+#define DT_REG		2
+#define DT_LNK		3
+#define DTYPE(de)	DT_UNKNOWN
+#endif
+
 /*
  * Basic data structures for the directory cache
  */
@@ -546,6 +561,9 @@ void set_alternate_index_output(const char *);
 
 extern int verify_index_checksum;
 extern int verify_ce_order;
+#ifdef __MVS__
+extern int ignore_file_tags;
+#endif
 
 #define MTIME_CHANGED	0x0001
 #define CTIME_CHANGED	0x0002
