diff --git a/combine-diff.c b/combine-diff.c
index b799862..ab6730c 100644
--- a/combine-diff.c
+++ b/combine-diff.c
@@ -19,6 +19,9 @@
 #include "userdiff.h"
 #include "oid-array.h"
 #include "revision.h"
+#ifdef __MVS__
+#include <zos.h>
+#endif
 
 static int compare_paths(const struct combine_diff_path *one,
 			  const struct diff_filespec *two)
@@ -1074,6 +1077,10 @@ static void show_patch_diff(struct combine_diff_path *elem, int num_parent,
 			ssize_t done;
 			int is_file, i;
 
+#ifdef __MVS__
+      __disableautocvt(fd);      
+#endif
+
 			elem->mode = canon_mode(st.st_mode);
 			/* if symlinks don't work, assume symlink if all parents
 			 * are symlinks
