diff --git i/config.c w/config.c
index b18b561..5fbd081 100644
--- i/config.c
+++ w/config.c
@@ -21,6 +21,7 @@
 #include "gettext.h"
 #include "git-zlib.h"
 #include "ident.h"
+#include "read-cache-ll.h"
 #include "repository.h"
 #include "lockfile.h"
 #include "mailmap.h"
@@ -1499,6 +1500,18 @@ static int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+ #ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+
+	if (!strcmp(var, "core.utf8ccsid")) {
+		utf8_ccsid = git_config_ulong(var, value, ctx->kvi);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
