diff --git a/config.c b/config.c
index 3846a37..4cb3e7c 100644
--- a/config.c
+++ b/config.c
@@ -39,6 +39,9 @@
 #include "worktree.h"
 #include "ws.h"
 #include "write-or-die.h"
+#ifdef __MVS__
+#include "read-cache-ll.h"
+#endif
 
 struct config_source {
 	struct config_source *prev;
@@ -1670,6 +1673,13 @@ static int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+ #ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
