diff --git a/config.c b/config.c
index b79baf8..57fe8b4 100644
--- a/config.c
+++ b/config.c
@@ -11,6 +11,7 @@
 #include "alloc.h"
 #include "date.h"
 #include "branch.h"
+#include "cache.h"
 #include "config.h"
 #include "convert.h"
 #include "environment.h"
@@ -1708,6 +1709,13 @@ static int git_default_core_config(const char *var, const char *value, void *cb)
 		return 0;
 	}
 
+#ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
