diff --git a/config.mak.uname b/config.mak.uname
index 5feb582..ee69ee6 100644
--- a/config.mak.uname
+++ b/config.mak.uname
@@ -655,15 +655,25 @@ ifeq ($(uname_S),NONSTOP_KERNEL)
 	SHELL_PATH = /usr/coreutils/bin/bash
 endif
 ifeq ($(uname_S),OS/390)
+	PERL_PATH = $(PERL5_HOME)/bin/perl
+	PERL_PATH_FOR_SCRIPTS = /bin/env perl
+	SHELL_PATH = $(BASH_HOME)/bin/bash
+	TEST_SHELL_PATH = $(SHELL_PATH)
+	SHELL_PATH_FOR_SCRIPTS = /bin/env bash
+	PYTHON_PATH := $(PYTHON_PATH)/python3 
 	NO_SYS_POLL_H = YesPlease
+	RUNTIME_PREFIX = YesPlease
 	NO_STRCASESTR = YesPlease
 	NO_REGEX = YesPlease
 	NO_MMAP = YesPlease
 	NO_NSEC = YesPlease
 	NO_STRLCPY = YesPlease
+	NO_MKDTEMP = YesPlease
 	NO_MEMMEM = YesPlease
 	NO_GECOS_IN_PWENT = YesPlease
 	HAVE_STRINGS_H = YesPlease
+	NEEDS_LIBICONV = YesPlease
+	ICONVDIR = $(LIBICONV_HOME)
 	NEEDS_MODE_TRANSLATION = YesPlease
 	HAVE_ZOS_GET_EXECUTABLE_PATH = YesPlease
 endif
