diff --git a/config.c b/config.c
index 3cfeb3d8bd..1968317fe4 100644
--- a/config.c
+++ b/config.c
@@ -6,6 +6,9 @@
  *
  */
 #include "git-compat-util.h"
+#ifdef __MVS__
+#include "read-cache-ll.h"
+#endif
 #include "abspath.h"
 #include "advice.h"
 #include "date.h"
@@ -1523,6 +1526,18 @@ static int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+ #ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+
+	if (!strcmp(var, "core.utf8ccsid")) {
+		utf8_ccsid = git_config_ulong(var, value, ctx->kvi);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
