diff --git a/config.c b/config.c
index 50f2d17b39..8cc8aab4d2 100644
--- a/config.c
+++ b/config.c
@@ -33,6 +33,9 @@
 #include "object-store-ll.h"
 #include "pager.h"
 #include "path.h"
+#ifdef __MVS__
+#include "read-cache-ll.h"
+#endif
 #include "utf8.h"
 #include "color.h"
 #include "refs.h"
@@ -1508,6 +1511,18 @@ static int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+ #ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+
+	if (!strcmp(var, "core.utf8ccsid")) {
+		utf8_ccsid = git_config_ulong(var, value, ctx->kvi);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
