diff --git i/date.c w/date.c
index 17a9507..be63eb6 100644
--- i/date.c
+++ w/date.c
@@ -6,12 +6,18 @@
 
 #define DISABLE_SIGN_COMPARE_WARNINGS
 
+//FIXME: localtime_r has an LE bug where it does not respect the TZ environment variable
+#define localtime_r localtime_r_replaced
 #include "git-compat-util.h"
 #include "date.h"
 #include "gettext.h"
 #include "pager.h"
 #include "strbuf.h"
+#undef localtime_r
 
+//FIXME: localtime_r has an LE bug where it does not respect the TZ environment variable
+// Use the EBCDIC symbol instead
+struct tm *localtime_r(const time_t *timep, struct tm *result) asm("@@LCLT@R"); 
 /*
  * This is like mktime, but without normalization of tm_wday and tm_yday.
  */
