diff --git a/diff.c b/diff.c
index a68ddd2..689cc66 100644
--- a/diff.c
+++ b/diff.c
@@ -4194,6 +4194,7 @@ int diff_populate_filespec(struct repository *r,
 	int check_binary = options ? options->check_binary : 0;
 	int err = 0;
 	int conv_flags = global_conv_flags_eol;
+  int autocvtToASCII;
 	/*
 	 * demote FAIL to WARN to allow inspecting the situation
 	 * instead of refusing.
@@ -4267,9 +4268,18 @@ int diff_populate_filespec(struct repository *r,
 			s->is_binary = 1;
 			return 0;
 		}
+#ifdef __MVS__
+    validate_codeset(r->index, s->path, &autocvtToASCII);
+#endif
 		fd = open(s->path, O_RDONLY);
 		if (fd < 0)
 			goto err_empty;
+
+#ifdef __MVS__
+    if (!autocvtToASCII)
+      __disableautocvt(fd);
+#endif
+
 		s->data = xmmap(NULL, s->size, PROT_READ, MAP_PRIVATE, fd, 0);
 		close(fd);
 		s->should_munmap = 1;
@@ -4374,6 +4384,10 @@ static void prep_temp_blob(struct index_state *istate,
 		blob = buf.buf;
 		size = buf.len;
 	}
+  
+#ifdef __MVS__
+  tag_file_as_working_tree_encoding(istate, path, temp->tempfile->fd);
+#endif
 	if (write_in_full(temp->tempfile->fd, blob, size) < 0 ||
 	    close_tempfile_gently(temp->tempfile))
 		die_errno("unable to write temp-file");
