diff --git a/entry.c b/entry.c
index 7817aee..d2ca664 100644
--- a/entry.c
+++ b/entry.c
@@ -139,6 +139,10 @@ static int streaming_write_entry(const struct cache_entry *ce, char *path,
 	if (fd < 0)
 		return -1;
 
+#ifdef __MVS__
+  	tag_file_as_working_tree_encoding(state->istate, path, fd);
+#endif
+
 	result |= odb_stream_blob_to_fd(the_repository->objects, fd, &ce->oid, filter, 1);
 	*fstat_done = fstat_checkout_output(fd, state, statbuf);
 	result |= close(fd);
@@ -385,6 +389,10 @@ static int write_entry(struct cache_entry *ce, char *path, struct conv_attrs *ca
 			return error_errno("unable to create file %s", path);
 		}
 
+#ifdef __MVS__
+    tag_file_as_working_tree_encoding(state->istate, path, fd);
+#endif
+
 		wrote = write_in_full(fd, new_blob, size);
 		if (!to_tempfile)
 			fstat_done = fstat_checkout_output(fd, state, &st);
@@ -487,6 +495,25 @@ int checkout_entry_ca(struct cache_entry *ce, struct conv_attrs *ca,
 	struct stat st;
 	struct conv_attrs ca_buf;
 
+
+#ifdef __MVS__
+  const char* git_utf8_ccsid_str = getenv("GIT_UTF8_CCSID");
+
+  if (git_utf8_ccsid_str != NULL) {
+      char* endptr;
+      errno = 0;
+      long conv = strtol(git_utf8_ccsid_str, &endptr, 10);
+
+      if (!conv) {
+          perror("Error converting GIT_UTF8_CCSID to short");
+      } else if (endptr == git_utf8_ccsid_str) {
+          fprintf(stderr, "No digits were found in GIT_UTF8_CCSID\n");
+      } else {
+          utf8_ccsid = conv;
+      }
+  }
+#endif
+
 	if (ce->ce_flags & CE_WT_REMOVE) {
 		if (topath)
 			/*
