diff --git a/environment.h b/environment.h
index 6f18286..fd0fccc 100644
--- a/environment.h
+++ b/environment.h
@@ -173,6 +173,17 @@ extern unsigned long pack_size_limit_cfg;
 
 extern int protect_hfs;
 extern int protect_ntfs;
+#ifdef __MVS__
+enum zos_encoding_fallback {
+	ZOS_ENCODING_FALLBACK_FAIL = 0,
+	ZOS_ENCODING_FALLBACK_TRANSLITERATE = 1
+};
+extern int ignore_file_tags;
+extern int utf8_ccsid;
+extern enum zos_encoding_fallback zos_encoding_fallback;
+extern int zos_encoding_fallback_is_env_set;
+#endif
+
 
 enum rebase_setup_type {
 	AUTOREBASE_NEVER = 0,
