diff --git a/environment.c b/environment.c
index ba2c601..22395a8 100644
--- a/environment.c
+++ b/environment.c
@@ -27,6 +27,7 @@
 #include "repository.h"
 #include "config.h"
 #include "refs.h"
+#include "read-cache-ll.h"
 #include "fmt-merge-msg.h"
 #include "commit.h"
 #include "strvec.h"
@@ -52,6 +53,12 @@ char *git_log_output_encoding;
 char *apply_default_whitespace;
 char *apply_default_ignorewhitespace;
 int fsync_object_files = -1;
+#ifdef __MVS__
+int ignore_file_tags = 0;
+enum zos_encoding_fallback zos_encoding_fallback = ZOS_ENCODING_FALLBACK_FAIL;
+int zos_encoding_fallback_is_env_set = 0;
+int utf8_ccsid = 1208;
+#endif
 int use_fsync = -1;
 enum fsync_method fsync_method = FSYNC_METHOD_DEFAULT;
 enum fsync_component fsync_components = FSYNC_COMPONENTS_DEFAULT;
@@ -398,6 +405,32 @@ int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+#ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+
+	if (!strcasecmp(var, "core.zosencodingfallback")) {
+		if (zos_encoding_fallback_is_env_set)
+			return 0;
+		if (!value)
+			return config_error_nonbool(var);
+		if (!strcasecmp(value, "transliterate"))
+			zos_encoding_fallback = ZOS_ENCODING_FALLBACK_TRANSLITERATE;
+		else if (!strcasecmp(value, "fail"))
+			zos_encoding_fallback = ZOS_ENCODING_FALLBACK_FAIL;
+		else
+			return error(_("unknown core.zosEncodingFallback value: %s"), value);
+		return 0;
+	}
+
+	if (!strcmp(var, "core.utf8ccsid")) {
+		utf8_ccsid = git_config_ulong(var, value, ctx->kvi);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
