diff --git a/environment.c b/environment.c
index 8ffbf92..bef0b5d 100644
--- a/environment.c
+++ b/environment.c
@@ -26,6 +26,7 @@
 #include "repository.h"
 #include "config.h"
 #include "refs.h"
+#include "read-cache-ll.h"
 #include "fmt-merge-msg.h"
 #include "commit.h"
 #include "strvec.h"
@@ -57,6 +58,10 @@ char *git_attributes_file;
 int zlib_compression_level = Z_BEST_SPEED;
 int pack_compression_level = Z_DEFAULT_COMPRESSION;
 int fsync_object_files = -1;
+#ifdef __MVS__
+int ignore_file_tags = 0;
+int utf8_ccsid = 1208;
+#endif
 int use_fsync = -1;
 enum fsync_method fsync_method = FSYNC_METHOD_DEFAULT;
 enum fsync_component fsync_components = FSYNC_COMPONENTS_DEFAULT;
@@ -406,6 +411,18 @@ int git_default_core_config(const char *var, const char *value,
 		return 0;
 	}
 
+#ifdef __MVS__
+	if (!strcmp(var, "core.ignorefiletags")) {
+		ignore_file_tags = git_config_bool(var, value);
+		return 0;
+	}
+
+	if (!strcmp(var, "core.utf8ccsid")) {
+		utf8_ccsid = git_config_ulong(var, value, ctx->kvi);
+		return 0;
+	}
+#endif
+
 	if (!strcmp(var, "core.safecrlf")) {
 		int eol_rndtrp_die;
 		if (value && !strcasecmp(value, "warn")) {
