diff --git a/git-compat-util.h b/git-compat-util.h
index 7c2a6538e5..1d9713a421 100644
--- a/git-compat-util.h
+++ b/git-compat-util.h
@@ -223,7 +223,11 @@ struct strbuf;
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <stddef.h>
+#define release stdlib_release
+#define fetch stdlib_fetch
 #include <stdlib.h>
+#undef fetch
+#undef release
 #include <stdarg.h>
 #include <stdbool.h>
 #include <string.h>
