diff --git a/builtin/hash-object.c b/builtin/hash-object.c
index 5d900a6..35c46f8 100644
--- a/builtin/hash-object.c
+++ b/builtin/hash-object.c
@@ -40,6 +40,10 @@ static void hash_object(const char *path, const char *type, const char *vpath,
 {
 	int fd;
 	fd = xopen(path, O_RDONLY);
+#ifdef __MVS__
+	if (fd >= 0 && __setfdbinary(fd))
+	  die_errno("Cannot set to binary '%s'", path);
+#endif  
 	hash_fd(fd, type, vpath, flags);
 }
 
