diff --git i/builtin/hash-object.c w/builtin/hash-object.c
index 6a99ec2..5bbfcad 100644
--- i/builtin/hash-object.c
+++ w/builtin/hash-object.c
@@ -39,6 +39,10 @@ static void hash_object(const char *path, const char *type, const char *vpath,
 {
 	int fd;
 	fd = xopen(path, O_RDONLY);
+#ifdef __MVS__
+	if (fd >= 0 && __setfdbinary(fd))
+	  die_errno("Cannot set to binary '%s'", path);
+#endif  
 	hash_fd(fd, type, vpath, flags);
 }
 
