diff --git i/builtin/hash-object.c w/builtin/hash-object.c
index a25f040..40ab473 100644
--- i/builtin/hash-object.c
+++ w/builtin/hash-object.c
@@ -62,6 +62,10 @@ static void hash_object(const char *path, const char *type, const char *vpath,
 {
 	int fd;
 	fd = xopen(path, O_RDONLY);
+#ifdef __MVS__
+	if (__setfdbinary(fd))
+	  die_errno("Cannot set to binary '%s'", path);
+#endif  
 	hash_fd(fd, type, vpath, flags, literally);
 }
 
