diff --git a/http.c b/http.c
index c8fc15aa11..d72d87f27e 100644
--- a/http.c
+++ b/http.c
@@ -1073,7 +1073,6 @@ static CURL *get_curl_handle(void)
 	if (ssl_cipherlist != NULL && *ssl_cipherlist)
 		curl_easy_setopt(result, CURLOPT_SSL_CIPHER_LIST,
 				ssl_cipherlist);
-
 	if (ssl_cert)
 		curl_easy_setopt(result, CURLOPT_SSLCERT, ssl_cert);
 	if (ssl_cert_type)
@@ -1336,6 +1335,10 @@ void http_init(struct remote *remote, const char *url, int proactive_auth)
 	set_from_env(&ssl_key_type, "GIT_SSL_KEY_TYPE");
 	set_from_env(&ssl_capath, "GIT_SSL_CAPATH");
 	set_from_env(&ssl_cainfo, "GIT_SSL_CAINFO");
+#ifdef __MVS__
+  if (!ssl_cainfo)
+	  set_from_env(&ssl_cainfo, "ZOPEN_GIT_SSL_CAINFO");
+#endif
 
 	set_from_env(&user_agent, "GIT_HTTP_USER_AGENT");
 
