diff --git a/imap-send.c b/imap-send.c
index af02c6a..7e1069e 100644
--- a/imap-send.c
+++ b/imap-send.c
@@ -254,11 +254,13 @@ static int verify_hostname(X509 *cert, const char *hostname)
 	if ((subj_alt_names = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL))) {
 		int num_subj_alt_names = sk_GENERAL_NAME_num(subj_alt_names);
 		for (i = 0; !found && i < num_subj_alt_names; i++) {
-			int ntype;
 			GENERAL_NAME *subj_alt_name = sk_GENERAL_NAME_value(subj_alt_names, i);
-			ASN1_STRING *subj_alt_str = GENERAL_NAME_get0_value(subj_alt_name, &ntype);
 
-			if (ntype == GEN_DNS && host_matches(hostname, subj_alt_str))
+			if (subj_alt_name->type == GEN_DNS &&
+			strlen((const char *)ASN1_STRING_get0_data(subj_alt_name->d.ia5)) ==
+			(size_t)ASN1_STRING_length(subj_alt_name->d.ia5) &&
+			host_matches(hostname,
+			(const char *)ASN1_STRING_get0_data(subj_alt_name->d.ia5)))
 				found = 1;
 		}
 		sk_GENERAL_NAME_pop_free(subj_alt_names, GENERAL_NAME_free);
