diff --git a/builtin/index-pack.c b/builtin/index-pack.c
index 0b62b2589f..91b3641342 100644
--- a/builtin/index-pack.c
+++ b/builtin/index-pack.c
@@ -484,9 +484,10 @@ static void *unpack_entry_data(off_t offset, unsigned long size,
 		oid = NULL;
 	if (type == OBJ_BLOB && size > big_file_threshold)
 		buf = fixed_buf;
-	else
+	else {
+		size += 1024;
 		buf = xmallocz(size);
-
+	}
 	memset(&stream, 0, sizeof(stream));
 	git_inflate_init(&stream);
 	stream.next_out = buf;
@@ -505,6 +506,11 @@ static void *unpack_entry_data(off_t offset, unsigned long size,
 			stream.avail_out = sizeof(fixed_buf);
 		}
 	} while (status == Z_OK);
+
+	 if (buf != fixed_buf) {
+		size -= 1024;
+	 }
+
 	if (stream.total_out != size || status != Z_STREAM_END)
 		bad_object(offset, _("inflate returned %d"), status);
 	git_inflate_end(&stream);
