diff --git i/compat/posix.h w/compat/posix.h
index 067a00f..1974083 100644
--- i/compat/posix.h
+++ w/compat/posix.h
@@ -284,6 +284,15 @@ int git_munmap(void *start, size_t length);
 #define MAP_FAILED ((void *)-1)
 #endif
 
+/* FIXME: utime is resulting in RACF messages, use git_utime to trace */
+#ifdef __MVS__
+#ifdef utime
+#undef utime
+#endif
+#define utime(path, times) git_utime(path, times)
+int git_utime(const char *path, const struct utimbuf *times);
+#endif
+
 #ifdef NEEDS_MODE_TRANSLATION
 #undef S_IFMT
 #undef S_IFREG
