diff --git a/repository.c b/repository.c
index c7e7521..a95755b 100644
--- a/repository.c
+++ b/repository.c
@@ -15,6 +15,9 @@
 #include "trace2.h"
 #include "promisor-remote.h"
 #include "refs.h"
+#ifdef __MVS__
+#include <zos.h>
+#endif
 
 /*
  * We do not define `USE_THE_REPOSITORY_VARIABLE` in this file because we do
@@ -446,7 +449,13 @@ int repo_hold_locked_index(struct repository *repo,
 			   struct lock_file *lf,
 			   int flags)
 {
+	int fd;
 	if (!repo->index_file)
 		BUG("the repo hasn't been setup");
-	return hold_lock_file_for_update(lf, repo->index_file, flags);
+	fd = hold_lock_file_for_update(lf, repo->index_file, flags);
+#ifdef __MVS__
+	if (fd >= 0)
+		__setfdbinary(fd);
+#endif
+	return fd;
 }
