diff --git a/builtin/cat-file.c b/builtin/cat-file.c
index d9fbad5..6f975e6 100644
--- a/builtin/cat-file.c
+++ b/builtin/cat-file.c
@@ -83,7 +83,8 @@ static int filter_object(const char *path, unsigned mode,
 		struct checkout_metadata meta;
 
 		init_checkout_metadata(&meta, NULL, NULL, oid);
-		if (convert_to_working_tree(the_repository->index, path, *buf, *size, &strbuf, &meta)) {
+		int ret = convert_to_working_tree(the_repository->index, path, *buf, *size, &strbuf, &meta);
+		if (ret > 0) {
 			free(*buf);
 			*size = strbuf.len;
 			*buf = strbuf_detach(&strbuf, NULL);
